src/AppInstallerRepositoryCore/Rest/RestSource.cpp

```diff
@@ -0,0 +1,262 @@
+                            && CaseInsensitiveEquals(versionInfo.GetChannel().ToString(), versionKey.Channel))
+                        {
+                            packageVersion = std::make_shared<PackageVersion>(source, m_package.packageInfo, versionInfo);
+                        }
+                    }
+                }
+                else if (versionKey.Version.empty() && versionKey.Channel.empty())
+                {
+                    return GetLatestAvailableVersion();
+                }
+                else if (versionKey.Version.empty())
+                {
+                    for (const auto& versionInfo : m_package.versions)
+                    {
+                        if (CaseInsensitiveEquals(versionInfo.GetChannel().ToString(), versionKey.Channel))
+                        {
+                            packageVersion = std::make_shared<PackageVersion>(source, m_package.packageInfo, versionInfo);
+                        }
+                    }
+                }
+                else if (versionKey.Channel.empty())
+                {
+                    for (const auto& versionInfo : m_package.versions)
+                    {
+                        if (CaseInsensitiveEquals(versionInfo.GetVersion().ToString(), versionKey.Version))
+                        {
+                            packageVersion = std::make_shared<PackageVersion>(source, m_package.packageInfo, versionInfo);
+                        }
+                    }
+                }
+
+                return packageVersion;
+            }
+
+            bool IsUpdateAvailable() const override
+            {
+                return false;
+            }
+        };
+    }
+
+    RestSource::RestSource(const SourceDetails& details, std::string identifier, RestClient&& restClient)
+        : m_details(details), m_identifier(std::move(identifier)), m_restClient(std::move(restClient))
+    {
+    }
+
+    const SourceDetails& RestSource::GetDetails() const
+    {
+        return m_details;
+    }
+
+    const RestClient& RestSource::GetRestClient() const
+    {
+        return m_restClient;
+    }
+
+    const std::string& RestSource::GetIdentifier() const
+    {
+        return m_identifier;
+    }
+
+    SearchResult RestSource::Search(const SearchRequest& request) const
+    {
+        // Note: Basic search functionality to fetch everything.
+        RestClient::SearchResult results = m_restClient.Search(request);
+        SearchResult searchResult;
+
+        std::shared_ptr<const RestSource> sharedThis = shared_from_this();
+        for (auto& result : results.Matches)
+        {
+            std::unique_ptr<IPackage> package = std::make_unique<AvailablePackage>(sharedThis, result);
+            PackageMatchFilter packageFilter({}, {}, {});
+
+            // Sort the versions
+            std::sort(result.versions.begin(), result.versions.end(),
+                [](const VersionAndChannel& a, const VersionAndChannel& b)

```

Comment:
This is the default < sort and shouldn't need to be specified.

Reply:
Wait, I see that you are only sorting based on version.  But why?

Reply:
Also, this sort is either useless due to a copy inside AvailablePackage or indicative of a lifetime bug there because it wasn't copied.

I would suggest just moving the sort inside of the object that is going to keep the copy (looks like PackageBase).

Reply:
you are right, fixed it to  sort by version and channel

---
In reply to: [583295822](https://github.com/microsoft/winget-cli/pull/765#discussion_r583295822) [](ancestors = 583295822)